app/boot_control/grub.py

```diff
@@ -0,0 +1,698 @@
+            logger.info(
+                "system doesn't use ota-partition mechanism to boot, "
+                "initializing ota-partition file..."
+            )
+            # NOTE: just copy but not cleanup the existed kernel/initrd files
+            shutil.copy(
+                self.boot_dir / cur_kernel,
+                self.active_ota_partition_folder,
+                follow_symlinks=True,
+            )
+            shutil.copy(
+                self.boot_dir / cur_initrd,
+                self.active_ota_partition_folder,
+                follow_symlinks=True,
+            )
+            self.reprepare_active_ota_partition_file(abort_on_standby_missed=False)
+
+        logger.info("ota-partition file initialized")
+
+    def grub_reboot_to_standby(self):
+        self.reprepare_standby_ota_partition_file()
+        idx, _ = GrubHelper.get_entry(
+            read_from_file(self.grub_file),
+            kernel_ver=GrubHelper.SUFFIX_OTA_STANDBY,
+        )
+        subprocess_call(f"grub-reboot {idx}", raise_exception=True)
+        logger.info(f"system will reboot to {self.standby_slot=}: boot entry {idx}")
+
+    def finalize_update_switch_boot(self):
+        self.reprepare_active_ota_partition_file(abort_on_standby_missed=True)
+
+
+class GrubController(
+    VersionControlMixin,
+    OTAStatusMixin,
+    PrepareMountMixin,
+    SlotInUseMixin,
+    BootControllerProtocol,
+):
+    def __init__(self) -> None:
+        try:
+            self._boot_control = _GrubControl()
+
+            # try to unmount standby dev if possible
+            CMDHelperFuncs.umount(self._boot_control.standby_root_dev)
+            self.standby_slot_mount_point = Path(cfg.MOUNT_POINT)
+            self.standby_slot_mount_point.mkdir(exist_ok=True)
+
+            ## ota-status dir
+            self.current_ota_status_dir = self._boot_control.active_ota_partition_folder
+            self.standby_ota_status_dir = (
+                self._boot_control.standby_ota_partition_folder
+            )
+
+            # refroot mount point
+            self.ref_slot_mount_point = Path(cfg.REF_ROOT_MOUNT_POINT)
+            # try to umount refroot mount point
+            CMDHelperFuncs.umount(self.ref_slot_mount_point)
+            if not os.path.isdir(self.ref_slot_mount_point):
+                os.mkdir(self.ref_slot_mount_point)
+
+            # init boot control
+            _ota_status = self._load_current_ota_status()
+            if _ota_status == OTAStatusEnum.UPDATING:
+                _ota_status = self._finalize_update()
+            elif _ota_status == OTAStatusEnum.ROLLBACKING:
+                _ota_status = self._finalize_rollback()
+            elif _ota_status == OTAStatusEnum.SUCCESS:
+                # need to check whether it is negative SUCCESS
+                current_slot = self._boot_control.active_slot
+                try:
+                    slot_in_use = self._load_current_slot_in_use()
```

Comment:
[IMPORTANT]
If OTA update is done by the current OTA client and then switch to this PR's OTA client.
Does this scenario work?

Reply:
Ah I will skip this check if the `slot_in_use` file is not presented to maintain compatibility. 
Maybe in the future I will revert the above workaround when all otaclients are new enough.

Reply:
@obi-t4 currently behavior is that if the `slot_in_use` file is not found, the otastatus will be set to `INITIALIZED` and ota-partition file will be regenerated. Is that OK for backward compatibility?

Reply:
(Previous comments are deleted as I am not thinking carefully.)
@obi-t4 if we are updating from old otaclient to new otaclient, then in the grub controller init process, it will do the `finalize_update`, but current `finalize_update` doesn't touch `slot_in_use` file. 
I will let `finalize_update` to write `slot_in_use` file, so the problem will be solved.

Thanks for pointing out this!

Reply:
> I will let finalize_update to write slot_in_use file, then the problem should be solved.

Can this operation handle an error case? I'm afraid it always becomes a success.

Reply:
Oh I update the implementation, now in the `OTAUpdater`, the `ota_status` file will always update to `FAILURE` when error happened during the running of `execute` method.

Reply:
So that restart `otaclient` after any exception will always result in ota_status == `FAILURE` now. 
Previously it will remain SUCCESS in some cases, that is an issue.